community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java

```diff
@@ -0,0 +1,196 @@
+     */
+    public AsyncEvents( Consumer<T> eventConsumer, Monitor monitor )
+    {
+        this.eventConsumer = eventConsumer;
+        this.monitor = monitor;
+        this.stack = shutdownSentinel;
+    }
+
+    @Override
+    public void send( T event )
+    {
+        AsyncEvent prev = STACK.getAndSet( this, event );
+        assert prev != null;
+        event.next = prev;
+        if ( prev == endSentinel )
+        {
+            LockSupport.unpark( runner );
+        }
+        else if ( prev == shutdownSentinel )
+        {
+            AsyncEvent events = STACK.getAndSet( this, shutdownSentinel );
+            process( events );
+        }
+    }
+
+    @Override
+    public void run()
+    {
+        assert runner == null : "A thread is already running " + runner;
+        runner = Thread.currentThread();
+
+        try
+        {
+            do
+            {
+                AsyncEvent events = STACK.getAndSet( this, endSentinel );
+                process( events );
+                if ( stack == endSentinel && !shutdown )
+                {
+                    LockSupport.park( this );
+                }
+            }
+            while ( !shutdown );
+
+            AsyncEvent events = STACK.getAndSet( this, shutdownSentinel );
+            process( events );
+        }
+        finally
+        {
+            runner = null;
+        }
+    }
+
+    private void process( AsyncEvent events )
+    {
+        events = reverseAndStripEndMark( events );
+
+        while ( events != null )
+        {
+            @SuppressWarnings( "unchecked" )
+            T event = (T) events;
+            eventConsumer.accept( event );
+            events = events.next;
+        }
+
+        endSentinel.next = null;
+        shutdownSentinel.next = null;
+    }
+
+    private AsyncEvent reverseAndStripEndMark( AsyncEvent events )
+    {
+        AsyncEvent result = null;
+        long count = 0;
+        while ( events != endSentinel && events != shutdownSentinel )
+        {
+            AsyncEvent tmp;
+            do
+            {
+                tmp = events.next;
+            }
+            while ( tmp == null );
+            events.next = result;
+            result = events;
+            events = tmp;
+            count++;
+        }
+        if ( count > 0 )
+        {
+            monitor.eventCount( count );
+        }
+        return result;
+    }
+
+    /**
+     * Initiate the shut down process of this {@code AsyncEvents} instance.
+     *
+     * This call does not block or otherwise wait for the background thread to terminate.
+     */
+    public void shutdown()
+    {
+        assert runner != null : "Already shut down";
+        shutdown = true;
+        LockSupport.unpark( runner );
+    }
+
+    public void awaitTermination() throws InterruptedException
+    {
+        while ( runner != null )
+        {
+            Thread.sleep( 10 );
+        }
```

Comment:
It's a nice thing to be able to do. There are other ways of achieving that, I'll implement such an other way :)